SpeechLLM LibriSpeech recipe

Adel-Moumen opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

-->

Not ready to be merged yet :) 

<!-- Does your PR introduce any breaking changes? If yes, please list them. -->

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

<!--

🎩 Magic happens when you code. Keep the spells flowing!

-->

pplantinga reviewed (commented):
Overall, looks like a good addition -- it will be nice to have a starting framework for doing SpeechLLMs in SpeechBrain!

From my first read, I guess I'm wondering if we want to try to support more than one task, even a simple second task such as keyword spotting, just to show how it can be done, as it seems like the main benefit of SpeechLLMs over traditional ASR is the fact that they can support multiple tasks.

TParcollet commented on recipes/LibriSpeech/ASR/transformer/README.md:
```diff
@@ -23,6 +22,20 @@ installed in your environment (see extra-requirements.txt)**
+
+| Release | Model | hyperparams file | Dev Clean WER | Test Clean WER | Test Other WER | HuggingFace link | Model link | GPUs |
+|:-------------:|:-------------:|:-------------:|:---------------------------:| :-----:| :-----:| :-----:| :--------:|
+| 29-11-25 | WavLM Large + SmolLM2 1.7B + LoRA | speechllm_ssl_feats.yaml | N/A | 3.17 | 6.83 | [HuggingFace](https://huggingface.co/speechbrain/asr-speechllm-librispeech) | - | 1xH100 40GB |
```
These are not really good numbers given the architecture imho, but ok.

pplantinga commented on speechbrain/nnet/activations.py:
```diff
@@ -169,3 +169,76 @@ def forward(self, x):
+        return F.gelu(x, approximate=self.approximate)
+
+
+class ReLU(torch.nn.Module):
```
Same here, I don't think this is needed.

pplantinga commented on speechbrain/nnet/schedulers.py:
```diff
@@ -300,7 +300,6 @@ class LinearWarmupScheduler:
     from the initial lr set in the optimizer to 0, after
     a warmup period during which it increases linearly
     from 0 to the initial lr set in the optimizer.
-    * Ge Li 2022
```
Did Ge Li not work on this? Why would we remove it?

pplantinga reviewed (commented):
Looks like its getting close, thanks Adel! Just a few small comments before merge.

Adel-Moumen commented on speechbrain/core.py:
```diff
@@ -457,6 +457,12 @@ def __init__(  # noqa: C901
+            if hasattr(self.modules, "module")
+            else self.modules
+        )
+
```
The point is to exactly remove the DDP extra layering (similar to here: https://github.com/karpathy/nanoGPT/blob/3adf61e154c3fe3fca428ad6bc3818b27a3b8291/train.py#L253). That way, we can systematically access to methods e.g. `get_input_embedding` which wouldn't be accessible easily (e.g. here: `self.raw_modules.get_input_embedding`) while before we would have to do something like `x = self.modules.module.get_input_embedding if 'module' in self.modules else ...` so you would still need to unwrap the DDP container to get target module for calling the function. Here, we just have a pointer that does that instead of doing it manually. I have seen some recipes that had to do a ugly 'if' and I think this would solve the problem. 

having a function that unwrap would work but I think this is simpler and more modular. Happy to consider something else if you think this is not the way to go.

Adel-Moumen commented:
> Looks like its getting close, thanks Adel! Just a few small comments before merge.

I think I addressed all the comments! Fixed the notebook and HDF5 (we were never passing the `compression` arg to the constructor) + removed the activations + added back authors.
